src/proof.rs

```diff
@@ -1,52 +1,483 @@
+            (
+                "merkleProofLength".to_string(),
+                vec![merkle_proof_length.to_string()],
+            ),
+            (
+                "merkleProofIndices".to_string(),
+                merkle_proof_indices.iter().map(|i| i.to_string()).collect(),
+            ),
+            (
+                "merkleProofSiblings".to_string(),
+                merkle_proof_siblings
+                    .iter()
+                    .map(|s| BigUint::from_bytes_le(s.to_vec().as_ref()).to_string())
+                    .collect(),
+            ),
+            ("scope".to_string(), vec![hash(scope_uint.clone())]),
+            ("message".to_string(), vec![hash(message_uint.clone())]),
+        ]);
+
+        let circom_proof = CircomProver::prove(
+            ProofLib::Arkworks,
+            WitnessFn::RustWitness(semaphore_witness),
+            serde_json::to_string(&inputs).unwrap(),
+            ZKEY_PATH.to_string(),
+        )?;
+
+        Ok(SemaphoreProof {
+            merkle_tree_depth,
+            merkle_tree_root: BigUint::from_bytes_le(merkle_proof.root.as_ref()),
+            message: message_uint,
+            nullifier: circom_proof.pub_inputs.0.get(1).unwrap().clone(),
+            scope: scope_uint,
+            points: Self::pack_groth16_proof(circom_proof.proof),
+        })
+    }
+
+    pub fn verify_proof(proof: SemaphoreProof) -> bool {
+        // check tree depth
+        if proof.merkle_tree_depth < MIN_TREE_DEPTH || proof.merkle_tree_depth > MAX_TREE_DEPTH {
+            panic!("The tree depth must be a number between and");
+        }
+
+        let scope = BigUint::from_str(hash(proof.scope).as_str()).unwrap();
+        let message = BigUint::from_str(hash(proof.message).as_str()).unwrap();
+        let pub_inputs = PublicInputs(vec![
+            proof.merkle_tree_root,
+            proof.nullifier,
+            message,
+            scope,
+        ]);
+        let p = CircomProof {
+            proof: Self::unpack_groth16_proof(proof.points),
+            pub_inputs,
+        };
+
+        CircomProver::verify(ProofLib::Arkworks, p, ZKEY_PATH.to_string()).unwrap()
     }
 
-    fn verify_proof(_proof: SemaphorProof) -> bool {
-        unimplemented!()
+    pub fn pack_groth16_proof(p: circom::Proof) -> PackedGroth16Proof {
+        [
+            p.a.x,
+            p.a.y,
+            p.b.x[0].clone(),
+            p.b.x[1].clone(),
+            p.b.y[0].clone(),
+            p.b.y[1].clone(),
+            p.c.x,
+            p.c.y,
+        ]
     }
 
-    fn pack_groth16_proof(_proof: Groth16Proof) -> PackedGroth16Proof {
-        unimplemented!()
+    pub fn unpack_groth16_proof(packed: PackedGroth16Proof) -> circom::Proof {
+        let a = G1 {
+            x: packed[0].clone(),
+            y: packed[1].clone(),
+            z: BigUint::one(),
+        };
+        let b = G2 {
+            x: [packed[2].clone(), packed[3].clone()],
+            y: [packed[4].clone(), packed[5].clone()],
```

Comment:
here is the same

```suggestion
            x: [packed[3].clone(), packed[2].clone()],
            y: [packed[5].clone(), packed[4].clone()],
```

https://github.com/privacy-scaling-explorations/zk-kit/blob/577fa314c5510cd347da0d7f2fcdc200326f60ef/packages/utils/src/proof-packing.ts#L40